modules/sdk/src/main/scala/org/tessellation/sdk/infrastructure/snapshot/MajorityPeerSelect.scala

```diff
@@ -50,28 +59,34 @@ object MajorityPeerSelect {
 
     def filterPeerList(peers: NonEmptyList[Peer]): F[NonEmptyList[Peer]] =
       peers
-        .traverse(snapshotClient.getLatestOrdinal(_))
-        .map {
-          _.groupBy(identity).maxBy { case (_, ordinals) => ordinals.size }
+        .parTraverseN(maxConcurrentPeerInquiries)(snapshotClient.getLatestOrdinal(_))
+        .map(_.groupBy(identity).maxBy { case (_, ordinals) => ordinals.size })
+        .map { case (majorityOrdinal, _) => majorityOrdinal }
+        .flatMap { majorityOrdinal =>
+          peers
+            .parTraverseN(maxConcurrentPeerInquiries)(peer => getPeerHash(peer, majorityOrdinal).attempt)
         }
-        .flatMap {
-          case (majorityOrdinal, _) =>
-            peers.traverse(snapshotClient.get(majorityOrdinal).run(_).flatMap(_.toHashed.map(_.hash)))
+        .flatMap { maybePeerSnapshots =>
+          MonadThrow[F].fromOption(
+            maybePeerSnapshots.collect { case Right(peerSnapshot) => peerSnapshot }.toNel,
+            NoPeersToSelect
+          )
         }
-        .map(_.zip(peers))
-        .map(_.groupMap { case (hash, _) => hash } { case (_, ps) => ps })
-        .map(_.maxBy { case (_, peers) => peers.size })
-        .map { case (_, peerCandidates) => peerCandidates }
+        .map(_.groupMap { case (_, hash) => hash } { case (peer, _) => peer })
+        .map(_.values.toList.sortWith(_.size > _.size))
+        .map(_.head)
 
     def getPeerSublist(peers: Set[Peer]): F[List[Peer]] = {
-      val maxSublistPercent = 0.25
-      val maxSublistSize = Math.max((peers.size * maxSublistPercent).toInt, 1)
+      val minSampleSize = Math.max((peers.size * maxPeerSampleRatio).toInt, 20)
 
-      Random[F].nextIntBounded(maxSublistSize).flatMap { peerCount =>
+      Random[F].nextIntBounded(1 + minSampleSize).flatMap { peerCount =>
```

Comment:
this does not fix the issue we discussed on slack. `nextIntBounded` may return `0`.
see above comment how I see that change

Reply:
@marcinwadon You're correct, that's my bad. Sorry @cngo-github  :(
The +1 belongs in the take:
```
 Random[F].nextIntBounded(minSampleSize).flatMap { peerCount =>
        Random[F]
          .shuffleList(peers.toList)
          .map(_.take(1 + peerCount))
      }
```
`Random[F].nextIntBounded(minSampleSize)` should give us a value in interval `[0, minSampleSize)`
so adding 1 in the take will give us `[1, minSampleSize]`

Reply:
.... which I just realized isn't really what we want.

Reply:
See my above comment.